### ISSUE TITLE
[Proposal] Versioning elements inside h5ad

### ISSUE THREAD

**Issue by ivirshup** [theislab/anndata#163]
I would like to version the elements inside an h5ad file (potentially also the zarr backend). This is coming from my work being able to write dataframes wherever we want on disk. In particular, I think there's a lot to gain from having dataframes work now, but I'm not sure the on-disk representation I currently have is optimal (random access might not be great). It is however, totally fine. By introducing versioning we'll be able keep encoding and decoding code simple, even if our on-disk representations change.

My proposal is, for any changes to the on-disk representation, to add a `encoding_type` and `encoding_version` field to the attrs of `h5ad` elements written from a single python object. For example, the `h5py.Dataset` would get attrs like:

```
{
  "encoding_type": "dataframe",
  "encoding_version": "v0.2.0"
}
```

This would make backwards compatibility much easier to maintain, since we won't have to "sniff" an `h5py` object to figure out how  to decode it. We'll just have instructions on what decoding function to call in the attrs. Specifics of the version for an encoding will are set by the functions which encode and decode them (so there spec becomes functional).

I'm not proposing we add this information to every piece of data right now, I think it's something we can add gradually as the encodings change.

**Comment by falexwolf** [comment]
I absolutely second this and had wanted to introduce something like this from the beginning. Just never had the time to think everything through...

Thank you, @ivirshup!

**Comment by ivirshup** [comment]
Do these names sound good? I figure it's important to stick with them once chosen.

**Comment by falexwolf** [comment]
Yes, the names sound good! And I agree, we shouldn't change them. 🙂
